build-coordinator/src/main/java/org/jboss/pnc/coordinator/builder/BuildQueue.java

```diff
@@ -0,0 +1,210 @@
+     */
+    public synchronized Optional<BuildTask> getTask(BuildConfigurationAudited buildConfigAudited) {
+        Optional<BuildTask> ready = readyTasks.stream().filter(bt -> bt.getBuildConfigurationAudited().equals(buildConfigAudited)).findAny();
+        Optional<BuildTask> waiting = waitingTasks.stream().filter(bt -> bt.getBuildConfigurationAudited().equals(buildConfigAudited)).findAny();
+        Optional<BuildTask> inProgress = tasksInProgress.stream().filter(bt -> bt.getBuildConfigurationAudited().equals(buildConfigAudited)).findAny();
+        return ready.isPresent() ? ready : waiting.isPresent() ? waiting : inProgress;
+    }
+
+    /**
+     * List all waiting, ready and in progress tasks
+     * @return list of all build tasks in the queue
+     */
+    public synchronized List<BuildTask> getSubmittedBuildTasks() {
+        ArrayList<BuildTask> result = new ArrayList<>();
+        result.addAll(waitingTasks);
+        result.addAll(readyTasks);
+        result.addAll(tasksInProgress);
+        return result;
+    }
+
+    private boolean isBuildAlreadySubmitted(BuildTask buildTask) {
+        return waitingTasks.contains(buildTask) || tasksInProgress.contains(buildTask);
+    }
+
+    private boolean rejectAlreadySubmitted(BuildTask buildTask) {
+        if (isBuildAlreadySubmitted(buildTask)) {
+            buildTask.setStatus(BuildCoordinationStatus.REJECTED);
+            buildTask.setStatusDescription("The configuration is already in the build queue.");
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    private List<BuildTask> extractReadyTasks() {
+        List<BuildTask> newReadyTasks = waitingTasks.stream()
+                .filter(BuildTask::readyToBuild)
+                .collect(Collectors.toList());
+        waitingTasks.removeAll(newReadyTasks);
+        return newReadyTasks;
+    }
+
+    private void startNext() {
+        while (tasksInProgress.size() < MAX_CONCURRENT_BUILDS) {
+            BuildTask task = readyTasks.poll();
+            if (task != null) {
+                tasksInProgress.add(task);
+                executorService.execute(() -> buildCoordinator.processBuildTask(task));
+            } else {
+                break;
+            }
+        }
+    }
+
+    private void addTask(BuildTask task) {
+        if (task.readyToBuild()) {
+            readyTasks.add(task);
+        } else {
+            waitingTasks.add(task);
+        }
+    }
+
+    public void setBuildCoordinator(BuildCoordinator buildCoordinator) {
+        this.buildCoordinator = buildCoordinator;
+    }
+
+    @Override
+    public String toString() {
+        return "BuildQueue{" +
+                "readyTasks=" + readyTasks +
+                ", waitingTasks=" + waitingTasks +
+                ", tasksInProgress=" + tasksInProgress +
+                ", taskSets=" + taskSets +
+                '}';
+    }
+
+    @Deprecated // used only in tests!
```

Comment:
actually, the method can be used not only in tests. It just checks if the queue is empty.

I think is worth having this check in tests. What would you say for removing Deprecated?